lib/transaction_ops.c

```diff
@@ -53,100 +53,120 @@
-		if (((pkg_repod = xbps_rpool_get_pkg(xhp, pkg)) == NULL) &&
-		    ((pkg_repod = xbps_rpool_get_virtualpkg(xhp, pkg)) == NULL)) {
+		*ttype = XBPS_TRANS_INSTALL;
+		if (((*pkg_repod = xbps_rpool_get_pkg(xhp, pkg)) == NULL) &&
+		    ((*pkg_repod = xbps_rpool_get_virtualpkg(xhp, pkg)) == NULL)) {
 			/* not found */
 			return ENOENT;
 		}
 	} else {
 		if (force) {
-			ttype = XBPS_TRANS_REINSTALL;
+			*ttype = XBPS_TRANS_REINSTALL;
 		} else {
-			ttype = XBPS_TRANS_UPDATE;
+			*ttype = XBPS_TRANS_UPDATE;
 		}
-		if (xbps_dictionary_get(pkg_pkgdb, "repolock")) {
+		if (xbps_dictionary_get(*pkg_pkgdb, "repolock")) {
 			struct xbps_repo *repo;
 			/* find update from repo */
-			xbps_dictionary_get_cstring_nocopy(pkg_pkgdb, "repository", &repoloc);
+			xbps_dictionary_get_cstring_nocopy(*pkg_pkgdb, "repository", &repoloc);
 			assert(repoloc);
 			if ((repo = xbps_regget_repo(xhp, repoloc)) == NULL) {
 				/* not found */
 				return ENOENT;
 			}
-			pkg_repod = xbps_repo_get_pkg(repo, pkg);
+			*pkg_repod = xbps_repo_get_pkg(repo, pkg);
 		} else {
 			/* find update from rpool */
-			pkg_repod = xbps_rpool_get_pkg(xhp, pkg);
+			*pkg_repod = xbps_rpool_get_pkg(xhp, pkg);
 		}
-		if (pkg_repod == NULL) {
+		if (*pkg_repod == NULL) {
 			/* not found */
 			return ENOENT;
 		}
 	}
 
-	xbps_dictionary_get_cstring_nocopy(pkg_repod, "pkgver", &repopkgver);
+	xbps_dictionary_get_cstring_nocopy(*pkg_repod, "pkgver", &repopkgver);
 
-	if (ttype == XBPS_TRANS_UPDATE) {
+	if (*ttype == XBPS_TRANS_UPDATE) {
 		/*
 		 * Compare installed version vs best pkg available in repos
 		 * for pkg updates.
 		 */
-		xbps_dictionary_get_cstring_nocopy(pkg_pkgdb,
+		xbps_dictionary_get_cstring_nocopy(*pkg_pkgdb,
 		    "pkgver", &instpkgver);
 		if (xbps_cmpver(repopkgver, instpkgver) <= 0 &&
-		    !xbps_pkg_reverts(pkg_repod, instpkgver)) {
-			xbps_dictionary_get_cstring_nocopy(pkg_repod,
+		    !xbps_pkg_reverts(*pkg_repod, instpkgver)) {
+			xbps_dictionary_get_cstring_nocopy(*pkg_repod,
 			    "repository", &repoloc);
 			xbps_dbg_printf(xhp, "[rpool] Skipping `%s' "
 			    "(installed: %s) from repository `%s'\n",
 			    repopkgver, instpkgver, repoloc);
 			return EEXIST;
 		}
-	} else if (ttype == XBPS_TRANS_REINSTALL) {
+	} else if (*ttype == XBPS_TRANS_REINSTALL) {
 		/*
 		 * For reinstallation check if installed version is less than
 		 * or equal to the pkg in repos, if true, continue with reinstallation;
 		 * otherwise perform an update.
 		 */
-		xbps_dictionary_get_cstring_nocopy(pkg_pkgdb, "pkgver", &instpkgver);
+		xbps_dictionary_get_cstring_nocopy(*pkg_pkgdb, "pkgver", &instpkgver);
 		if (xbps_cmpver(repopkgver, instpkgver) == 1) {
-			ttype = XBPS_TRANS_UPDATE;
+			*ttype = XBPS_TRANS_UPDATE;
 		}
 	}
 
+	return 0;
+}
+
+static int
+trans_find_pkg(struct xbps_handle *xhp, const char *pkg, bool force)
+{
+	xbps_dictionary_t pkg_pkgdb = NULL, pkg_repod = NULL;
+	xbps_object_t obj;
+	xbps_array_t pkgs;
+	pkg_state_t state = 0;
```

Comment:
`state` doesn't need to be initialized either, but idk if the compiler might complain with `maybe-uninitialized`; we'd have to make sure `xbps_pkg_state_installed` always initializes correctly, which seems like the right thing to do.